src/utils/FlowUtils.ts

```diff
@@ -31,61 +32,47 @@ export const V3_DEPOSIT_COMPARISON_KEYS = [
 
 export function filledSameDeposit(fillA: Fill, fillB: Fill): boolean {
   if (isV2Fill(fillA) && isV2Fill(fillB)) {
-    return (
-      fillA.depositId === fillB.depositId &&
-      fillA.originChainId === fillB.originChainId &&
-      fillA.amount.eq(fillB.amount) &&
-      fillA.destinationChainId === fillB.destinationChainId &&
-      fillA.relayerFeePct.eq(fillB.relayerFeePct) &&
-      fillA.recipient === fillB.recipient &&
-      fillA.depositor === fillB.depositor &&
-      fillA.message === fillB.message
-    );
+    return getV2RelayHash(fillA) === getV2RelayHash(fillB);
   } else if (isV3Fill(fillA) && isV3Fill(fillB)) {
-    return (
-      fillA.depositId === fillB.depositId &&
-      fillA.originChainId === fillB.originChainId &&
-      fillA.destinationChainId === fillB.destinationChainId &&
-      fillA.recipient === fillB.recipient &&
-      fillA.depositor === fillB.depositor &&
-      fillA.inputToken === fillB.inputToken &&
-      fillA.outputToken === fillB.outputToken &&
-      fillA.message === fillB.message &&
-      fillA.inputAmount.eq(fillB.inputAmount) &&
-      fillA.outputAmount.eq(fillB.outputAmount) &&
-      fillA.fillDeadline === fillB.fillDeadline &&
-      fillA.exclusivityDeadline === fillB.exclusivityDeadline &&
-      fillA.exclusiveRelayer === fillB.exclusiveRelayer
-    );
+    const { destinationChainId: chainA } = fillA;
+    const { destinationChainId: chainB } = fillB;
+    return chainA === chainB && getV3RelayHash(fillA, chainA) === getV3RelayHash(fillB, chainB);
   }
 
   return false;
 }
 
-// Ensure that each deposit element is included with the same value in the fill. This includes all elements defined
-// by the depositor as well as the realizedLpFeePct and the destinationToken, which are pulled from other clients.
-export function validateFillForDeposit(fill: Fill, deposit?: Deposit, fillFieldsToIgnore: string[] = []): boolean {
+export function validateFillForDeposit(
+  relayData: RelayData & { destinationChainId: number }, // V2Deposit, V3Fill, SlowFillRequest...
+  deposit?: Deposit,
+  fillFieldsToIgnore: string[] = []
+): boolean {
   if (deposit === undefined) {
     return false;
   }
 
-  if (isV2Deposit(deposit) && isV2Fill(fill)) {
-    return V2_DEPOSIT_COMPARISON_KEYS.every((key) => {
-      if (fillFieldsToIgnore.includes(key)) {
-        return true;
-      }
-      return fill[key] !== undefined && fill[key].toString() === deposit[key]?.toString();
-    });
+  return isV2RelayData(relayData)
+    ? validateV2FillForDeposit(relayData, deposit, fillFieldsToIgnore)
+    : validateV3FillForDeposit(relayData, deposit);
+}
+
+function validateV2FillForDeposit(fill: V2RelayData, deposit: Deposit, fillFieldsToIgnore: string[] = []): boolean {
+  if (!isV2Deposit(deposit)) {
+    return false;
   }
 
-  if (isV3Deposit(deposit) && isV3Fill(fill)) {
-    return V3_DEPOSIT_COMPARISON_KEYS.every((key) => {
```

Comment:
Note - technically we could remove V3_DEPOSIT_COMPARISON_KEYS in this PR. I've left that definition in because I foresee reusing it in this PR (https://github.com/across-protocol/sdk-v2/pull/493) in order to produce the underlying reason in the event of a mismatch. For example, if the hashes mismatch, then we proceed to iterate over the keys to identify which field(s) disagree.